packages/financial-templates-lib/src/clients/OptimisticOracleClient.js

```diff
@@ -73,54 +100,118 @@ class OptimisticOracleClient {
-        // or alternatively search for ProposedPrice events.
-        this.priceProposals.push(requestData);
-      }
     });
 
+    // Filter proposals based on their expiration timestamp:
+    const isExpired = proposal => {
+      return Number(proposal.expirationTimestamp) <= Number(currentTime);
+    };
+    this.expiredProposals = undisputedProposals.filter(proposal => isExpired(proposal));
+    this.undisputedProposals = undisputedProposals.filter(proposal => !isExpired(proposal));
+
+    // Store disputes that can be settled:
+    let resolvedPrices = await Promise.all(
+      disputeEvents.map(async disputeEvent => {
+        try {
+          // When someone disputes an OO proposal, the OO requests a price to the DVM with a re-formatted
+          // ancillary data packet that includes the original requester's information:
+          const stampedAncillaryData = await this.oracle.methods
+            .stampAncillaryData(
+              disputeEvent.returnValues.ancillaryData ? disputeEvent.returnValues.ancillaryData : "0x",
+              disputeEvent.returnValues.requester
+            )
+            .call();
+
+          // getPrice will return null or revert if there is no price available,
+          // in which case we'll ignore this dispute.
+          let resolvedPrice = revertWrapper(
+            await this.voting.methods
+              .getPrice(disputeEvent.returnValues.identifier, disputeEvent.returnValues.timestamp, stampedAncillaryData)
+              .call({
+                from: this.oracle.options.address
+              })
+          );
+          if (resolvedPrice !== null) {
+            return disputeEvent;
+          }
+        } catch (error) {
+          // No resolved price available, do nothing.
+        }
+      })
+    );
+    // Remove undefined entries, marking disputes that did not have resolved prices
+    resolvedPrices = resolvedPrices.filter(event => event !== undefined);
+
+    // Filter out disputes that were already settled and reformat data.
+    resolvedPrices = await Promise.all(
+      resolvedPrices.map(async event => {
+        const state = await this.oracle.methods
+          .getState(
+            event.returnValues.requester,
+            event.returnValues.identifier,
+            event.returnValues.timestamp,
+            event.returnValues.ancillaryData ? event.returnValues.ancillaryData : "0x"
+          )
+          .call();
+
+        // For unsettled disputes, reformat the data:
+        if (state !== OptimisticOracleRequestStatesEnum.SETTLED) {
+          return {
+            requester: event.returnValues.requester,
+            proposer: event.returnValues.proposer,
+            disputer: event.returnValues.disputer,
+            identifier: this.hexToUtf8(event.returnValues.identifier),
+            timestamp: event.returnValues.timestamp
+          };
+        }
+      })
+    );
+    resolvedPrices = resolvedPrices.filter(event => event !== undefined);
```

Comment:
Same here on potentially chaining to avoid reassignment (totally optional/up to you).

Reply:
Implemented [here](https://github.com/UMAprotocol/protocol/pull/2395/commits/9a44bad632db1cfe586127e4c6a9346c0cb30aa0)